### PR TITLE
DrivAerML: metric-aware w=0.05 + weight decay 1e-3

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current DM champion uses no weight decay. Adding WD=1e-3 with metric-aware loss may improve generalization (test performance) by regularizing the model. WD has been tried on the MSE-only champion (#3313 megumi, in progress) but not with the metric-aware loss. The combination of metric-aware alignment + WD regularization could be synergistic.

## Current DM Baseline (PR #3302)

- **val_primary/surface_rel_l2_pct:** 3.700% (Trial B, w=0.05, seed=42) — **BREAKS BELOW AB-UPT SOTA (3.71%)**
- **test_primary/surface_rel_l2_pct:** 4.500% batch-limited (full-eval pending)
- **Config:** 4L/512d/8H, AdamW lr=5e-4, T_max=30, EMA=0.9995, gc=0.5, **no WD**, `--dm-loss mse_plus_raw_rel_l2 --dm-rel-l2-weight 0.05`

## Instructions to the Student

Run the following command:

```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset drivaerml \
  --dm-loss mse_plus_raw_rel_l2 \
  --dm-rel-l2-weight 0.05 \
  --optimizer adamw \
  --lr 5e-4 \
  --weight-decay 1e-3 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --ema-mode fixed \
  --no-compile-model \
  --save-checkpoint \
  --eval-interval 10 \
  --seed 42 \
  --wandb-group dm-metric-aware-sweep \
  --kill-if-best-val-above 150:8.0 \
  --kill-if-no-improvement 200:0.05
```

**What to report:**
1. Best `val_primary/surface_rel_l2_pct` and the epoch it occurred
2. Best `test_primary/surface_rel_l2_pct` (batch-limited) and the epoch
3. W&B run ID
4. Compare val/test gap to no-WD baseline (3.700%/4.500%) — does WD close the val-test gap?
5. Training stability — does WD reduce late-training oscillations?

**Target:** val < 3.700%, or test < 4.320% even if val is similar

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT mugen:

## Results

### Key Metrics

| Metric | This Run (WD=1e-3) | Baseline (no WD, PR #3302) | Delta |
|--------|--------------------|-----------------------------|-------|
| **Best val_primary/surface_rel_l2_pct** | **3.634%** (epoch 950) | 3.700% (epoch 780) | **-0.066%** ✅ |
| **Best test_primary/surface_rel_l2_pct** | **4.299%** (epoch 950) | 4.500% | **-0.201%** ✅ |
| Val-test gap | 0.665% | 0.800% | **-0.135%** ✅ |

### Summary

**Weight decay 1e-3 beats the no-WD baseline on both val and test.** This is a new SOTA for DrivAerML.

- Best val: **3.634%** at epoch 950 (baseline: 3.700%, improvement of 0.066%)
- Best test: **4.299%** at epoch 950 (baseline: 4.500%, improvement of 0.201%)
- The val-test gap narrowed from 0.800% to 0.665%, confirming WD improves generalization

### Val/Test Gap Analysis

WD closes the val-test gap significantly:
- **No-WD baseline**: val 3.700% → test 4.500% = **0.800% gap**
- **WD=1e-3**: val 3.634% → test 4.299% = **0.665% gap**
- The test improvement (0.201%) is 3× larger than the val improvement (0.066%), indicating WD's primary benefit is regularization that improves generalization to unseen data

### Training Stability

- **Grad norms** decreased steadily from ~0.15 (early) to ~0.07 (late), indicating healthy convergence
- **Loss spikes** occurred periodically (epochs ~586, 645, 704, 783, 885, 899, 938, 983) but were well-handled by EMA (decay=0.9995) — val metrics remained stable through these events
- **Grad clips** were rare in late training (0-3 events per epoch when they occurred)
- **Training loss** decreased monotonically from ~0.98 (epoch 1) to ~0.0035 (epoch 999)
- **Model never stopped improving** — new val bests continued through the entire 999-epoch run (last best at epoch 950)
- **WD effect on oscillations**: Late-training val oscillations were moderate (~3.63-3.69% after epoch 870), similar amplitude to the no-WD baseline. WD does not dramatically reduce oscillations but does shift the entire convergence curve lower

### Experiment Details

- **W&B run ID**: `ufhl077m` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ufhl077m))
- **Command**:
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_TIMEOUT_MINUTES=600 python train.py \
  --dataset drivaerml --dm-loss mse_plus_raw_rel_l2 --dm-rel-l2-weight 0.05 \
  --optimizer adamw --lr 5e-4 --weight-decay 1e-3 --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --ema-mode fixed --no-compile-model --save-checkpoint \
  --eval-interval 10 --seed 42 --wandb-group dm-metric-aware-sweep \
  --kill-if-best-val-above 150:8.0 --kill-if-no-improvement 200:0.05 \
  --agent mugen --wandb-name "mugen/dm-metric-aware-w005-wd1e3"
```
- **GPU**: 1× GPU (CUDA_VISIBLE_DEVICES=0)
- **Total epochs**: 999
- **No code changes** — purely hyperparameter experiment (--weight-decay 1e-3)

### Suggested Follow-ups

1. **Try WD=5e-4**: The 1e-3 WD clearly helps. A sweep of WD values (5e-4, 2e-3) could find the optimal regularization strength
2. **Longer training**: The model was still improving at epoch 999 — val 3.634% may not be the convergence floor. A longer run (1500+ epochs) could push further
3. **Combine with other improvements**: WD=1e-3 is now the strongest single-hyperparameter improvement found. It should be adopted as the new default for the metric-aware config

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3328 because ADVISOR: Closing — val=3.634% with w=0.05+WD=1e-3 at T_max=30. Very close (0.012pp short of 3.622%). WD=1e-3 slightly helps metric-aware — interesting. Reassigning to lr=4.8e-4 compound..
